### PR TITLE
Add global timeout flag

### PR DESCRIPTION
**Pull request by juanvallejo**
**Release note**:

``` release-note
Add a new global option "--request-timeout" to the `kubectl` client
```

UPSTREAM: https://github.com/kubernetes/client-go/pull/10

This patch adds a global timeout flag (viewable with `kubectl -h`) with
a default value of `0s` (meaning no timeout).

The timeout value is added to the default http client, so that zero
values and default behavior are enforced by the client.

Adding a global timeout ensures that user-made scripts won't hang for an
indefinite amount of time while performing remote calls (right now, remote
calls are re-tried up to 10 times when each attempt fails, however, there is
no option to set a timeout in order to prevent any of these 10 attempts from
hanging indefinitely).

**Example**

```
$ kubectl get pods # no timeout flag set - default to 0s (which means no
timeout)
NAME                      READY     STATUS    RESTARTS   AGE
docker-registry-1-h7etw   1/1       Running   1          2h
router-1-uv0f9            1/1       Running   1          2h

$ kubectl get pods --request-timeout=0 # zero means no timeout no timeout flag set
NAME                      READY     STATUS    RESTARTS   AGE
docker-registry-1-h7etw   1/1       Running   1          2h
router-1-uv0f9            1/1       Running   1          2h

$kubectl get pods --request-timeout=1ms
Unable to connect to the server: net/http: request canceled while
waiting for connection (Client.Timeout exceeded while awaiting headers)
```

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33958)

### CODE REVIEW

File: pkg/kubectl/cmd/rollingupdate.go [hunk 1]
```diff
@@ -91,7 +91,7 @@ func NewCmdRollingUpdate(f *cmdutil.Factory, out io.Writer) *cobra.Command {
 	}
 	cmd.Flags().Duration("update-period", updatePeriod, `Time to wait between updating pods. Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".`)
 	cmd.Flags().Duration("poll-interval", pollInterval, `Time delay between polling for replication controller status after the update. Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".`)
-	cmd.Flags().Duration("timeout", timeout, `Max time to wait for a replication controller to update before giving up. Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".`)
```
Comment (deads2k) [inline pkg/kubectl/cmd/rollingupdate.go]: You cannot change these flag names without breaking existing scripts.

Comment (juanvallejo) [inline pkg/kubectl/cmd/rollingupdate.go]: reverted local flag names back to `timeout`. Updated global flag to `--request-timeout` https://github.com/kubernetes/kubernetes/pull/33958/commits/396551ba76ea7b3c13d875b09c955f2c6762671f

**Comment by kargakis** [comment]
cc: @kubernetes/kubectl

**Comment by juanvallejo** [comment]
@deads2k or @fabianofranz could you take a look please?

File: pkg/client/unversioned/clientcmd/overrides.go [hunk 2]
```diff
@@ -121,6 +138,7 @@ const (
 	FlagImpersonate  = "as"
 	FlagUsername     = "username"
 	FlagPassword     = "password"
+	FlagTimeout      = "request-timeout"
```
Comment (deads2k) [inline pkg/client/unversioned/clientcmd/overrides.go]: @kubernetes/kubectl Any comments on the flag name?  `--timeout` was already take by commands that use reapers for a different meaning.

**Comment by deads2k** [comment]
Squash and lgtm.

I'll leave it open a day for comments on the flag name.

**Comment by deads2k** [comment]
Oh, and test-cmd test.  A `kubectl get -w --request-timeout` ought to return, right?

**Comment by juanvallejo** [comment]
@deads2k 
> A kubectl get -w --request-timeout ought to return, right?

Yes, when using the default `0` value, it behaves as usual (watches a resource until a user terminates the process).

When used with a non-zero `request-timeout` value, it does timeout after the time specified:
```
$ oc get pod zookeeper-1 -w --request-timeout=10s
NAME          READY     STATUS    RESTARTS   AGE
zookeeper-1   1/1       Running   2          6h

```

After `10s`, the complete output from the command above is:
```
NAME          READY     STATUS    RESTARTS   AGE
zookeeper-1   1/1       Running   2          6h
E1005 15:50:25.125415   11319 streamwatcher.go:109] Unable to decode an event from the watch stream: net/http: request canceled (Client.Timeout exceeded while reading body)
```

**Comment by deads2k** [comment]
> Yes, when using the default 0 value, it behaves as usual (watches a resource until a user terminates the process).

Right, just pointed that out for your test-cmd.sh test.

File: pkg/client/unversioned/clientcmd/overrides.go (hunk 2 above)
Comment (AdoHe) [inline pkg/client/unversioned/clientcmd/overrides.go]: `request-timeout` sounds good to me.

**Comment by AdoHe** [comment]
@juanvallejo any priority between local timeout flag and global flag?

**Comment by soltysh** [comment]
> @juanvallejo any priority between local timeout flag and global flag?

I guess, since `request-timeout` is applied at transport layer it'll have priority over local timeout, which is usually applied to a bigger chunk of operation (ie. for reapers it's applied while waiting for scale down operation).

The changes LGTM, but since this is a new feature, release not is needed @juanvallejo,

**Comment by juanvallejo** [comment]
Thanks for the feedback, release note added!

**Comment by soltysh** [comment]
If there are no objections I'm gonna tag it for merge later today.

**Comment by soltysh** [comment]
@k8s-bot ok to test

**Comment by soltysh** [comment]
Flake #32430

@k8s-bot node e2e test this

File: pkg/client/unversioned/clientcmd/overrides.go [hunk 3]
```diff
@@ -105,6 +108,20 @@ func (f FlagInfo) BindBoolFlag(flags *pflag.FlagSet, target *bool) {
 	}
 }
 
+// BindDurationFlag binds the flag based on the provided info.  If LongName == "", nothing is registered
+func (f FlagInfo) BindDurationFlag(flags *pflag.FlagSet, target *time.Duration) {
```
Comment (fabianofranz) [inline pkg/client/unversioned/clientcmd/overrides.go]: I still believe `--request-timeout` should support values without units, which defaults to seconds like every other timeout flag out there. Like in `ssh`, `timeout`, `scp`, `curl`, and so on. 

You'd probably have to use string and parse it by yourself like in [here](https://play.golang.org/p/LgfIAgEvDC).

**Review by fabianofranz — CHANGES_REQUESTED** [review]

File: pkg/client/unversioned/clientcmd/overrides.go (hunk 3 above)
Comment (fabianofranz) [inline pkg/client/unversioned/clientcmd/overrides.go]: Like `--request-timeout=30` is the first thing I'll try.

Comment (juanvallejo) [inline pkg/client/unversioned/clientcmd/overrides.go]: Done! Thanks for providing the example

**Comment by fabianofranz** [comment]
LGTM

**Comment by soltysh** [comment]
@juanvallejo can you look into those test failures and link to flakes?

File: staging/src/k8s.io/client-go/1.5/rest/config.go [hunk 4]
```diff
@@ -109,6 +110,9 @@ type Config struct {
 	// Rate limiter for limiting connections to the master from this client. If present overwrites QPS/Burst
 	RateLimiter flowcontrol.RateLimiter
 
+	// The maximum length of time to wait before giving up on a server request. A value of zero means no timeout.
+	Timeout time.Duration
+
```
Comment (soltysh) [inline staging/src/k8s.io/client-go/1.5/rest/config.go]: What about actually setting this field like you do in the core k8s code? I'd expect this patch looks identical to the other `config.go` file.

**Review by soltysh — CHANGES_REQUESTED** [review]
@juanvallejo can you link the actual client-go PR here, as well

File: staging/src/k8s.io/client-go/1.5/rest/config.go (hunk 4 above)
Comment (juanvallejo) [inline staging/src/k8s.io/client-go/1.5/rest/config.go]: Definitely, will do right now

**Comment by juanvallejo** [comment]
Linking bump of `k8s.io/client-go/1.5/rest/config.go`: https://github.com/kubernetes/client-go/pull/10

**Comment by deads2k** [comment]
@juanvallejo `test-cmd` doesn't often flake.  Try it locally.

**Comment by juanvallejo** [comment]
@soltysh Finally getting tests to pass, however I'm not sure if the `GKE smoke e2e` is a flake or not

> What about actually setting this field like you do in the core k8s code? I'd expect this patch looks identical to the other config.go file.

I went ahead and added the rest of the changes from the core k8s code to the v1.5 config patchhttps://github.com/kubernetes/kubernetes/pull/33958/commits/2e0865356b524d62a787c48721a04ca8a485b91b

**Comment by juanvallejo** [comment]
gke e2e test flaked on https://github.com/kubernetes/kubernetes/issues/33388

**Comment by juanvallejo** [comment]
tests flaking on https://github.com/kubernetes/kubernetes/issues/32428

**Comment by soltysh** [comment]
@k8s-bot test this issue #32428

**Review by soltysh — APPROVED** [review]

**Comment by juanvallejo** [comment]
@soltysh or @deads2k PTAL

**Comment by AdoHe** [comment]
LGTM.

**Comment by deads2k** [comment]
@caesarxuchao is the "bump(k8s.io/client-go): add Timeout field to 1.5 restclient" commit well formed?  I haven't looked at how staging is supposed to be populated.

**Comment by deads2k** [comment]
squash your non-bump commits, get confirmation you've done the right thing for staging, and this lgtm.

**Comment by caesarxuchao** [comment]
@juanvallejo did you manually copy the exact changes in the first commit to the staging area? If so, I'm fine with doing it manually this time. (sorry for the trouble we are causing by the restructuring of client-go)

Also, could keep the commits separate for this PR? Just in case we need to revert the changes to the staging area.

**Comment by juanvallejo** [comment]
@caesarxuchao 
>  did you manually copy the exact changes in the first commit to the staging area?

Yes, this was done manually. I'll keep both commits separate. Thanks!

**Comment by caesarxuchao** [comment]
LGTM. Thanks!

**Pull request merged**